### PR TITLE
fix: oauth1 signing for url encoded content

### LINKED ISSUE #1: OAuth 1.0: support URL form encoded parameters
_From [yan...@google.com](https://code.google.com/u/111815493751521097215/) on May 11, 2011 10:11:50_

Copied from: https://code.google.com/p/google-api-java-client/issues/detail?id=95 Describe the problem.

Form encoded POST parameters are part of the signature base string and must be signed.  Looking at OAuthAuthorizationHeaderIntercepter it is clear that the interceptor does not do this: it simply passes request method and URL to the OAuthParameters object.  Consequently, the computed signature is wrong and the POST request fails.
This comes up when, e.g. tweeting using the Twitter API:  their API expects a form encoded "status" parameter. GET requests using the OAuthAuthorizationHeaderIntercepter work fine.

How would you expect it to be fixed?

One possible fix would be to provide a second OAuthParameters#computeSignature that accepts a collection of additional parameters that form part of the signature base string.  The OAuthAuthorizationHeaderIntercepter could then check the HttpRequest's method and in the case of a form encoded POST request pass the additional parameters to the #computeSignature method.

Comment 1 by project member yan...@google.com, Jan 8, 2011

Thanks for reporting this issue.  This is actually a known limitation: http://javadoc.google-api-java-client.googlecode.com/hg/1.2.2-alpha/com/google/api/client/auth/oauth/package-summary.html But thanks for bringing it to my attention.  We do want this OAuth library to work properly with the Twitter API.

_Original issue: http://code.google.com/p/google-oauth-java-client/issues/detail?id=1_

### PR DESCRIPTION
**Pull request by cmunden**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-oauth-java-client/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1 ☕️

### CODE REVIEW

File: google-oauth-client/src/main/java/com/google/api/client/auth/oauth/OAuthParameters.java [hunk 1]
```diff
@@ -14,16 +14,14 @@
 
 package com.google.api.client.auth.oauth;
 
-import com.google.api.client.http.GenericUrl;
-import com.google.api.client.http.HttpExecuteInterceptor;
-import com.google.api.client.http.HttpRequest;
-import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.*;
```
Comment (elharo) [inline google-oauth-client/src/main/java/com/google/api/client/auth/oauth/OAuthParameters.java:17]: Google doesn't use wildcard imports

**Review by elharo — COMMENTED** [review]
an issue explaining the bug would help.

File: google-oauth-client/src/main/java/com/google/api/client/auth/oauth/OAuthParameters.java (hunk 1 above)
Comment (cmunden) [reply google-oauth-client/src/main/java/com/google/api/client/auth/oauth/OAuthParameters.java]: I have removed the wildcard import in my most recent commit.  Regarding an issue to explain the bug, a pre-existing issue was referenced in the pull request description.

Fixes #1

**Comment by elharo** [comment]
@silvolu The referenced issue is pretty old. Is this still something we want to do?

File: google-oauth-client/src/test/java/com/google/api/client/auth/oauth/OAuthParametersTest.java [hunk 2]
```diff
@@ -72,6 +80,22 @@ public void testSignature() throws GeneralSecurityException {
     assertEquals("GET&https%3A%2F%2Fexample.local&foo%3Dbar%26oauth_signature_method%3Dmock", parameters.signature);
   }
 
+  public void testSignatureWithUrlEncodedContent() throws IOException {
+    OAuthParameters parameters = new OAuthParameters();
+    parameters.signer = new MockSigner();
+
+    GenericUrl url = new GenericUrl("https://example.local?foo=bar");
+    Map<String, Object> contentParameters = Collections.singletonMap("this", (Object) "that");
```
[the remark this answers is not in the document]
Comment (cmunden) [reply google-oauth-client/src/test/java/com/google/api/client/auth/oauth/OAuthParametersTest.java:88]: The cast to Object is necessary in order to create the contentParameters templated to a map of type Map<String, Object>.  Without the explicit cast to Object, the singletonMap method will template to a map of type Map<String, String>.

**Comment by vladokrsymphony** [comment]
Hello, any plans soon to merge this PR?

**Review by elharo — APPROVED** [review]

**Comment by cmunden** [comment]
@vladokrsymphony  I recommend that your report your additional issue regarding non ascii characters in the Issues section.  Just go [here](https://github.com/googleapis/google-oauth-java-client/issues/new/choose) and choose "Get Started" in the Bug Report area.

Since this PR is now merged and your bug is of a slightly different nature, then it can be tracked and not missed.  Also remember to provide some good step-by-step instructions on how to recreate the issue.

Happy coding 👍

**Pull request merged**